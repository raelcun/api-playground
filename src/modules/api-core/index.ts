export * from './types'
export * from './response'
