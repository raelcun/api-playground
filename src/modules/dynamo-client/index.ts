import { DocumentClient } from 'aws-sdk/clients/dynamodb'

import { createDynamoClient } from '@lib/dynamo-client'
import { getConfig } from '@modules/config'

export const getDynamoClient = createDynamoClient(
  new DocumentClient({
    region: getConfig().aws.region,
    credentials: getConfig().aws.credentials,
  }),
)
