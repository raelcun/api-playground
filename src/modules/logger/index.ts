export * from './logger'
export { middleware as logRequest } from './logRequestMiddleware'
