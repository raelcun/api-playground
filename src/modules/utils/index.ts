import { KoaContext } from '@modules/api-core'
import { createMockContext, Options } from '@shopify/jest-koa-mocks'

export const createKoaContext = <T>(
  options?: Options<Record<string, unknown>, T> & { requestBody?: T } & {
    body?: string | Record<string, unknown>
  },
): KoaContext<T> => {
  const context = createMockContext(options) as KoaContext<T>
  if (options !== undefined && options.body !== undefined) {
    context.body = options.body
  }
  return context
}

export const createMockNext = (impl = () => {}): jest.Mock<Promise<void>> =>
  jest.fn(() => {
    impl()
    return Promise.resolve()
  })
