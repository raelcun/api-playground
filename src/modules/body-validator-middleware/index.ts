import { withValidatedBody as innerMiddleware } from '@lib/body-validator-middleware'
import { getSystemLogger } from '@modules/logger'

export const withValidatedBody = innerMiddleware(getSystemLogger)
