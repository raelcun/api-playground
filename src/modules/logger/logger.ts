import { createLogger, Logger } from '@lib/logger'
import { getConfig } from '@modules/config'

export const getLogger = createLogger(() => getConfig().logging)

export const getSystemLogger = (): Logger => getLogger('system')
