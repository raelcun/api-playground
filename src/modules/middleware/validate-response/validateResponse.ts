import * as t from 'io-ts'
import { reporter } from 'io-ts-reporters'
import { left } from 'fp-ts/lib/Either'
import HttpStatus from 'http-status-codes'
import { Middleware } from '../../../types'
import { getSystemLogger } from '../../logger'
import { getConfig } from '../../../config'

export const validateResponseInner = (shouldError: boolean) => <T>(
  type: t.Type<T, unknown>,
): Middleware<T> => async (ctx, next) => {
  type.decode(ctx.body).fold(
    errors => {
      getSystemLogger().warn('failed to validate response\n' + reporter(left(errors)))
      if (shouldError) {
        ctx.status = HttpStatus.INTERNAL_SERVER_ERROR
        ctx.body = {
          errors: reporter(left(errors)),
        }
      } else {
        next()
      }
    },
    () => {
      next()
    },
  )
}

export const validateResponse = validateResponseInner(getConfig().isProduction)
