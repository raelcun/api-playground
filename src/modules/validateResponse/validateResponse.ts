import * as t from 'io-ts'
import { either as E } from 'fp-ts'
import HttpStatus from 'http-status-codes'
import { pipe } from 'fp-ts/lib/pipeable'
import { Lazy } from 'fp-ts/lib/function'
import { LoggerFactory } from 'modules/logger/types'
import { decode, logErrors, mapErrorCode } from 'utils'
import { Err } from 'modules/error/types'
import { Middleware } from 'koa'
import { getConfig } from 'config'
import { getSystemLogger } from 'modules/logger'

export const validateResponseInternal = (createLogger: LoggerFactory) => <T>(
  type: t.Type<T, unknown>,
) => (response: unknown): E.Either<Err<'RESPONSE_VALIDATION_ERROR'>, T> =>
  pipe(
    decode(type, response),
    mapErrorCode('RESPONSE_VALIDATION_ERROR'),
    logErrors(createLogger(), 'warn'),
  )

export const validateResponseMiddleware = (
  createLogger: LoggerFactory,
  shouldError: Lazy<boolean>,
) => <T>(type: t.Type<T, unknown>): Middleware => async (ctx, next) => {
  pipe(
    validateResponseInternal(createLogger)(type)(ctx.body),
    E.fold(
      () => {
        if (shouldError()) {
          ctx.status = HttpStatus.INTERNAL_SERVER_ERROR
        } else {
          next()
        }
      },
      () => {
        next()
      },
    ),
  )
}

/* istanbul ignore next */
export const validateResponse = validateResponseMiddleware(
  getSystemLogger,
  () => !getConfig().isProduction,
)
