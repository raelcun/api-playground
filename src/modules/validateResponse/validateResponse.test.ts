import * as t from 'io-ts'
import { createKoaContext, createMockNext } from 'utils'
import { getSystemLogger } from 'modules/logger'
import { validateResponseMiddleware } from './validateResponse'

describe('validate-response', () => {
  test.each([false, true])(
    'should log error when validation fails and shouldThrow is %s',
    async shouldThrow => {
      const logger = getSystemLogger()
      const warnLoggerSpy = jest.spyOn(logger, 'warn')
      const next = createMockNext()
      const ctx = createKoaContext({ body: 'foo' })

      await validateResponseMiddleware(
        () => logger,
        () => shouldThrow,
      )(t.number)(ctx, next)

      expect(warnLoggerSpy.mock.calls).toMatchSnapshot()
    },
  )

  test('should call next on invalid response when shouldThrow is false', async () => {
    const next = createMockNext()
    const ctx = createKoaContext({ requestBody: 'foo' })

    await validateResponseMiddleware(getSystemLogger, () => false)(t.number)(ctx, next)

    expect(next).toHaveBeenCalledTimes(1)
  })

  test('should not call next on invalid response when shouldThrow is true', async () => {
    const next = createMockNext()
    const ctx = createKoaContext({ requestBody: 'foo' })

    await validateResponseMiddleware(getSystemLogger, () => true)(t.number)(ctx, next)

    expect(next).not.toHaveBeenCalled()
  })

  test('should have 500 status code when invalid response and shouldThrow is true', async () => {
    const ctx = createKoaContext({ requestBody: 'foo' })

    await validateResponseMiddleware(getSystemLogger, () => true)(t.number)(ctx, createMockNext())

    expect(ctx.status).toEqual(500)
  })

  test('should call next on valid response', async () => {
    const next = createMockNext()
    const ctx = createKoaContext({ body: '5' })

    await validateResponseMiddleware(getSystemLogger, () => true)(t.string)(ctx, next)

    expect(next).toHaveBeenCalledTimes(1)
  })
})
