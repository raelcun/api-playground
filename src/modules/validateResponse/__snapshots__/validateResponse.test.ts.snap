// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate-response should log error when validation fails and shouldThrow is false 1`] = `
Array [
  Array [
    Object {
      "code": "RESPONSE_VALIDATION_ERROR",
      "message": "Expecting number but instead got: \\"foo\\".",
    },
  ],
]
`;

exports[`validate-response should log error when validation fails and shouldThrow is true 1`] = `
Array [
  Array [
    Object {
      "code": "RESPONSE_VALIDATION_ERROR",
      "message": "Expecting number but instead got: \\"foo\\".",
    },
  ],
]
`;
